src/sync/rwlock.rs

```diff
@@ -254,31 +269,35 @@ impl<T> Drop for RwLockReadGuard<'_, T> {
     fn drop(&mut self) {
         self.inner = None;
 
-        if ExecutionState::should_stop() {
-            return;
-        }
-
-        // Unblock every thread waiting on this lock. The scheduler will choose one of them to win
-        // the race to this lock, and that thread will re-block all the losers.
-        let me = ExecutionState::me();
         let mut state = self.state.borrow_mut();
+
+        trace!(
+            holder = ?state.holder,
+            waiting_readers = ?state.waiting_readers,
+            waiting_writers = ?state.waiting_writers,
+            "releasing Read lock on rwlock {:p}",
+            self.state
+        );
+
         match &mut state.holder {
             RwLockHolder::Read(readers) => {
-                readers.remove(me);
+                let was_reader = readers.remove(self.me);
+                assert!(was_reader);
                 if readers.is_empty() {
                     state.holder = RwLockHolder::None;
                 }
             }
             _ => panic!("exiting a reader but rwlock is in the wrong state"),
         }
-        RwLock::<T>::unblock_waiters(&*state, me, RwLockType::Read);
-        trace!(
-            holder = ?state.holder,
-            waiting_readers = ?state.waiting_readers,
-            waiting_writers = ?state.waiting_writers,
-            "released Read lock on rwlock {:p}",
-            self.state
-        );
+
+        if ExecutionState::should_stop() {
```

Comment:
Can you say why this check had to be moved later?

Reply:
This is the check that bails out if we're currently panicking. The idea is to try to return the lock to some sort of usable state before bailing out, so that code further along the unwinding path can re-acquire the lock.

We originally added this check to avoid trying to access `ExecutionState` during a panic. Now that I think about it, there might be some cleaner way to do this where we get rid of `should_stop` and have these drop handlers instead be more cautious about whether they can access `ExecutionState` (with a method like `ExecutionState::try_with` or something). I'll think about that some more, but in the meantime I think change this is still an improvement on the current situation.